bookkeeper-common/src/main/java/org/apache/bookkeeper/common/component/Lifecycle.java

```diff
@@ -0,0 +1,196 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.bookkeeper.common.component;
+
+/**
+ * Lifecycle state. Allows the following transitions:
+ *
+ * <ul>
+ * <li>INITIALIZED -&gt; STARTED, STOPPED, CLOSED</li>
+ * <li>STARTED     -&gt; STOPPED</li>
+ * <li>STOPPED     -&gt; STARTED, CLOSED</li>
+ * <li>CLOSED      -&gt; </li>
+ * </ul>
+ *
+ * <p>Also allows to stay in the same state. For example, when calling stop on a component, the
+ * following logic can be applied:
+ *
+ * <pre>
+ * public void stop() {
+ *  if (!lifecycleState.moveToStopped()) {
+ *      return;
+ *  }
+ * // continue with stop logic
+ * }
+ * </pre>
+ *
+ * <p>Note, closed is only allowed to be called when stopped, so make sure to stop the component first.
+ * Here is how the logic can be applied:
+ *
+ * <pre>
+ * public void close() {
+ *  if (lifecycleState.started()) {
+ *      stop();
+ *  }
+ *  if (!lifecycleState.moveToClosed()) {
+ *      return;
+ *  }
+ *  // perform close logic here
+ * }
+ * </pre>
+ */
+public class Lifecycle {
```

Comment:
It seems to me that there is not concurrency control in this class, why ?

Reply:
this class usually used during statup and teardown. it is usually done by one thread. there is really need a lot of concurrency control attentions. and the caller can take care of this concurrency control in the place when it is used.

Reply:
Ok

Reply:
Ok